File: src/Scanner.Core/Rules/RuleLoader.cs
using System.Text.Json;

namespace Scanner.Core.Rules;

/// <summary>
/// Loads rule definitions from per-team directories and validates them.
/// </summary>
public sealed class RuleLoader
{
    private readonly RuleSchemaValidator _validator = new();

    /// <summary>
    /// Loads all rule definitions from the provided rules root.
    /// </summary>
    /// <param name="rulesRoot">Root directory containing per-team rule folders.</param>
    /// <returns>Rules grouped by team.</returns>
    /// <exception cref="ArgumentException">Thrown when the rules root is missing.</exception>
    /// <exception cref="DirectoryNotFoundException">Thrown when the rules directory is missing.</exception>
    public IReadOnlyList<TeamRules> LoadRules(string rulesRoot)
    {
        if (string.IsNullOrWhiteSpace(rulesRoot))
        {
            throw new ArgumentException("Rules root is required.", nameof(rulesRoot));
        }

        if (!Directory.Exists(rulesRoot))
        {
            throw new DirectoryNotFoundException($"Rules directory not found: {rulesRoot}");
        }

        var teamRules = new List<TeamRules>();
        foreach (var teamDirectory in Directory.EnumerateDirectories(rulesRoot))
        {
            var teamName = Path.GetFileName(teamDirectory);
            var rules = new List<RuleDefinition>();
            foreach (var file in Directory.EnumerateFiles(teamDirectory))
            {
                var extension = Path.GetExtension(file);
                if (!extension.Equals(".json", StringComparison.OrdinalIgnoreCase)
                    && !extension.Equals(".yml", StringComparison.OrdinalIgnoreCase)
                    && !extension.Equals(".yaml", StringComparison.OrdinalIgnoreCase))
                {
                    continue;
                }

                var rule = LoadRule(file);
                rules.Add(rule);
            }

            teamRules.Add(new TeamRules(teamName, rules));
        }

        return teamRules;
    }

    /// <summary>
    /// Loads and validates rules under the provided root.
    /// </summary>
    /// <param name="rulesRoot">Root directory containing per-team rule folders.</param>
    /// <returns>The validation result including any errors.</returns>
    public RuleValidationResult ValidateRules(string rulesRoot)
    {
        var teamRules = LoadRules(rulesRoot);
        var errors = new List<RuleValidationError>();
        foreach (var team in teamRules)
        {
            foreach (var rule in team.Rules)
            {
                var ruleErrors = _validator.Validate(rule);
                foreach (var error in ruleErrors)
                {
                    errors.Add(new RuleValidationError(team.TeamName, rule.Id, error));
                }
            }
        }

        return new RuleValidationResult(teamRules, errors);
    }

    /// <summary>
    /// Loads a single rule definition from a JSON or YAML file.
    /// </summary>
    /// <param name="path">The file path to the rule definition.</param>
    /// <returns>The parsed rule definition.</returns>
    /// <exception cref="InvalidDataException">Thrown when the rule file cannot be parsed.</exception>
    public RuleDefinition LoadRule(string path)
    {
        var extension = Path.GetExtension(path);
        if (extension.Equals(".json", StringComparison.OrdinalIgnoreCase))
        {
            var json = File.ReadAllText(path);
            var rule = JsonSerializer.Deserialize<RuleDefinition>(json, new JsonSerializerOptions
            {
                PropertyNameCaseInsensitive = true
            });
            if (rule == null)
            {
                throw new InvalidDataException($"Rule file {path} is empty or invalid.");
            }

            return rule;
        }

        if (extension.Equals(".yml", StringComparison.OrdinalIgnoreCase) || extension.Equals(".yaml", StringComparison.OrdinalIgnoreCase))
        {
            return ParseSimpleYamlRule(File.ReadAllLines(path));
        }

        throw new InvalidDataException($"Unsupported rule file format: {path}");
    }

    private static RuleDefinition ParseSimpleYamlRule(IEnumerable<string> lines)
    {
        var values = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        foreach (var line in lines)
        {
            var trimmed = line.Trim();
            if (string.IsNullOrWhiteSpace(trimmed) || trimmed.StartsWith("#", StringComparison.Ordinal))
            {
                continue;
            }

            var index = trimmed.IndexOf(':');
            if (index <= 0)
            {
                continue;
            }

            var key = trimmed[..index].Trim();
            var value = trimmed[(index + 1)..].Trim().Trim('"');
            values[key] = value;
        }

        values.TryGetValue("id", out var id);
        values.TryGetValue("description", out var description);
        values.TryGetValue("severity", out var severity);
        values.TryGetValue("checkId", out var checkId);
        values.TryGetValue("appliesTo", out var appliesTo);
        values.TryGetValue("recommendation", out var recommendation);

        return new RuleDefinition(
            id ?? string.Empty,
            description ?? string.Empty,
            severity ?? string.Empty,
            checkId ?? string.Empty,
            string.IsNullOrWhiteSpace(appliesTo) ? null : appliesTo,
            string.IsNullOrWhiteSpace(recommendation) ? null : recommendation);
    }
}

/// <summary>
/// Captures a validation error for a specific rule.
/// </summary>
/// <param name="Team">The owning team.</param>
/// <param name="RuleId">The rule identifier.</param>
/// <param name="Message">The validation error message.</param>
public sealed record RuleValidationError(string Team, string RuleId, string Message);

/// <summary>
/// Summarizes rule validation results.
/// </summary>
/// <param name="Teams">The loaded team rules.</param>
/// <param name="Errors">The validation errors.</param>
public sealed record RuleValidationResult(IReadOnlyList<TeamRules> Teams, IReadOnlyList<RuleValidationError> Errors)
{
    /// <summary>
    /// Gets a value indicating whether the validation completed with no errors.
    /// </summary>
    public bool IsValid => Errors.Count == 0;
}

File: node_engine/src/rules/RuleLoader.js
const fs = require("fs");
const path = require("path");
const { RuleSchemaValidator } = require("./RuleSchemaValidator");

const AllowedProperties = new Set([
  "id",
  "description",
  "severity",
  "checkId",
  "appliesTo",
  "recommendation",
  "wcagCriteria",
  "problemTags"
]);

const RequiredProperties = ["id", "description", "severity", "checkId"];

class RuleValidationResult {
  constructor(teams, errors) {
    this.teams = teams;
    this.errors = errors;
  }

  get isValid() {
    return this.errors.length === 0;
  }
}

class RuleLoader {
  constructor() {
    this.validator = new RuleSchemaValidator();
  }

  loadRules(rulesRoot) {
    if (!rulesRoot || !rulesRoot.trim()) {
      throw new Error("Rules root is required.");
    }

    if (!fs.existsSync(rulesRoot)) {
      throw new Error(`Rules directory not found: ${rulesRoot}`);
    }

    const teamRules = [];
    for (const teamDirectory of fs.readdirSync(rulesRoot, { withFileTypes: true })) {
      if (!teamDirectory.isDirectory()) {
        continue;
      }

      const teamName = teamDirectory.name;
      const rules = [];
      const teamPath = path.join(rulesRoot, teamName);
      for (const file of fs.readdirSync(teamPath)) {
        const extension = path.extname(file).toLowerCase();
        if (![".json", ".yml", ".yaml"].includes(extension)) {
          continue;
        }

        const result = this.loadRuleFile(path.join(teamPath, file), false);
        rules.push(result.rule);
      }

      teamRules.push({ teamName, rules });
    }

    return teamRules;
  }

  validateRules(rulesRoot) {
    if (!rulesRoot || !rulesRoot.trim()) {
      throw new Error("Rules root is required.");
    }

    if (!fs.existsSync(rulesRoot)) {
      throw new Error(`Rules directory not found: ${rulesRoot}`);
    }

    const teamRules = [];
    const errors = [];

    for (const teamDirectory of fs.readdirSync(rulesRoot, { withFileTypes: true })) {
      if (!teamDirectory.isDirectory()) {
        continue;
      }

      const teamName = teamDirectory.name;
      const rules = [];
      const teamPath = path.join(rulesRoot, teamName);
      for (const file of fs.readdirSync(teamPath)) {
        const extension = path.extname(file).toLowerCase();
        if (![".json", ".yml", ".yaml"].includes(extension)) {
          continue;
        }

        const result = this.loadRuleFile(path.join(teamPath, file), true);
        rules.push(result.rule);
        for (const error of result.errors) {
          errors.push({ team: teamName, ruleId: result.ruleIdForError, message: error });
        }

        const ruleErrors = this.validator.validate(result.rule);
        for (const error of ruleErrors) {
          errors.push({ team: teamName, ruleId: result.ruleIdForError, message: error });
        }
      }

      teamRules.push({ teamName, rules });
    }

    return new RuleValidationResult(teamRules, errors);
  }

  loadRule(pathToRule) {
    return this.loadRuleFile(pathToRule, false, true).rule;
  }

  loadRuleFile(pathToRule, captureErrors, allowJsonExceptions = false) {
    try {
      const extension = path.extname(pathToRule).toLowerCase();
      if (extension === ".json") {
        const json = fs.readFileSync(pathToRule, "utf-8");
        if (captureErrors) {
          try {
            return parseJsonRule(json, pathToRule);
          } catch (error) {
            if (error instanceof SyntaxError) {
              return {
                rule: emptyRule(),
                errors: [`Rule file ${pathToRule} contains empty or invalid JSON.`],
                ruleIdForError: path.basename(pathToRule, extension),
                hasParseError: true
              };
            }
          }
        }

        const result = parseJsonRule(json, pathToRule);
        if (result.hasParseError) {
          throw new Error(result.errors.join(" "));
        }

        return result;
      }

      if (extension === ".yml" || extension === ".yaml") {
        const lines = fs.readFileSync(pathToRule, "utf-8").split(/\r?\n/);
        const result = parseSimpleYamlRule(lines, pathToRule);
        if (!captureErrors && result.hasParseError) {
          throw new Error(result.errors.join(" "));
        }

        return result;
      }

      throw new Error(`Unsupported rule file format: ${pathToRule}`);
    } catch (error) {
      if (error instanceof SyntaxError && !allowJsonExceptions) {
        throw new Error(`Rule file ${pathToRule} contains empty or invalid JSON.`);
      }

      if (captureErrors) {
        return {
          rule: emptyRule(),
          errors: [error.message],
          ruleIdForError: path.basename(pathToRule, path.extname(pathToRule)),
          hasParseError: true
        };
      }

      throw error;
    }
  }
}

const emptyRule = () => ({
  id: "",
  description: "",
  severity: "",
  checkId: "",
  appliesTo: null,
  recommendation: null,
  wcagCriteria: null,
  problemTags: null
});

const parseJsonRule = (json, pathToRule) => {
  const errors = [];
  const document = JSON.parse(json);

  if (!document || typeof document !== "object" || Array.isArray(document)) {
    errors.push("Rule definition must be a JSON object.");
    return {
      rule: emptyRule(),
      errors,
      ruleIdForError: path.basename(pathToRule, path.extname(pathToRule)),
      hasParseError: true
    };
  }

  const values = {};
  for (const [name, value] of Object.entries(document)) {
    if (!AllowedProperties.has(name)) {
      errors.push(`Unknown property '${name}'.`);
      continue;
    }

    if (typeof value === "string") {
      values[name] = value;
    } else {
      errors.push(`Property '${name}' must be a string.`);
      values[name] = null;
    }
  }

  for (const required of RequiredProperties) {
    if (!(required in values)) {
      errors.push(`Missing required property '${required}'.`);
    }
  }

  const rule = {
    id: values.id ?? "",
    description: values.description ?? "",
    severity: values.severity ?? "",
    checkId: values.checkId ?? "",
    appliesTo: normalizeOptional(values.appliesTo),
    recommendation: normalizeOptional(values.recommendation),
    wcagCriteria: normalizeOptional(values.wcagCriteria),
    problemTags: normalizeOptional(values.problemTags)
  };

  const ruleId = rule.id || path.basename(pathToRule, path.extname(pathToRule));
  return {
    rule,
    errors,
    ruleIdForError: ruleId,
    hasParseError: errors.length > 0
  };
};

const parseSimpleYamlRule = (lines, pathToRule) => {
  const values = {};
  const errors = [];

  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) {
      continue;
    }

    const index = trimmed.indexOf(":");
    if (index <= 0) {
      continue;
    }

    const key = trimmed.slice(0, index).trim();
    const value = trimmed.slice(index + 1).trim().replace(/^"|"$/g, "");
    if (!AllowedProperties.has(key)) {
      errors.push(`Unknown property '${key}'.`);
      continue;
    }

    values[key] = value;
  }

  for (const required of RequiredProperties) {
    if (!(required in values)) {
      errors.push(`Missing required property '${required}'.`);
    }
  }

  const rule = {
    id: values.id ?? "",
    description: values.description ?? "",
    severity: values.severity ?? "",
    checkId: values.checkId ?? "",
    appliesTo: normalizeOptional(values.appliesTo),
    recommendation: normalizeOptional(values.recommendation),
    wcagCriteria: normalizeOptional(values.wcagCriteria),
    problemTags: normalizeOptional(values.problemTags)
  };

  const ruleId = rule.id || path.basename(pathToRule, path.extname(pathToRule));
  return {
    rule,
    errors,
    ruleIdForError: ruleId,
    hasParseError: errors.length > 0
  };
};

const normalizeOptional = (value) => {
  if (value == null) {
    return null;
  }

  const trimmed = value.trim();
  return trimmed ? trimmed : null;
};

module.exports = { RuleLoader, RuleValidationResult, parseJsonRule, parseSimpleYamlRule };
